### PR TITLE
Fix error handling in parsing 'using' statement

### LINKED ISSUE #2992: Possible bug embedding `using module` inside a method call
_From @rjmholt on January 11, 2017 2:12_

Was deleting a `using module Foo` statement inside a `[powershell]::AddStatement(...)` method call. Screenshot attached:
psbug

_Copied from original issue: lzybkr/PSReadLine#489_

### PR DESCRIPTION
**Pull request by iSazonov**
1. Fix error handling to close #2992.
2. Fix error reporting.
3. Add tests.

The PR overlaps #3005.
/cc rjmholt Sorry, I didn't see your PR because you don't announce your work in #2992

### CODE REVIEW

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 1]
```diff
@@ -4414,8 +4414,10 @@ private StatementAst UsingStatementRule(Token usingToken)
             {
                 case TokenKind.EndOfInput:
                 case TokenKind.NewLine:
-                // Example, using module ,FooBar
+                // Example: 'using module ,FooBar'
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/parser/Parser.cs]: This comment should be improved (I wrote this code, but had to think again why this case was necessary).

`GetCommandArgument` will successfully return an argument for a unary array argument - but we don't want to allow that syntax with a using statement.

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 2]
```diff
@@ -4414,8 +4414,10 @@ private StatementAst UsingStatementRule(Token usingToken)
...
                 case TokenKind.Comma:
+                // Example: 'using module ;'
+                case TokenKind.Semi:
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/parser/Parser.cs]: I don't think `Semi` needs a comment, it's rather obvious, probably even more obvious if you move the case to be next to `NewLine`.

File: test/powershell/Language/Classes/scripting.Classes.using.tests.ps1 [hunk 3]
```diff
@@ -261,6 +261,23 @@ New-Object Foo
...
+            $err = Get-ParseResults "using module )"
+            $err.Count | Should Be 1
+            $err[0].ErrorId | Should Be 'InvalidValueForUsingItemName'
+        }
+
+        It "report an error when a value after 'using module' is not a valid module name" {
+            $err = Get-ParseResults "using module 123"
+            $err.Count | Should Be 1
+            $err[0].ErrorId | Should Be 'InvalidValueForUsingItemName'
+        }
```
Comment (lzybkr) [inline test/powershell/Language/Classes/scripting.Classes.using.tests.ps1]: You did not add any tests for a semi-colon terminated using statement.

**Review by lzybkr — CHANGES_REQUESTED** [review]
Thanks - the changes look pretty good, just some minor feedback.

File: src/System.Management.Automation/engine/parser/Parser.cs (hunk 1 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: Fixed.

File: src/System.Management.Automation/engine/parser/Parser.cs (hunk 2 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: Fixed.

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 4]
```diff
@@ -4426,11 +4428,14 @@ private StatementAst UsingStatementRule(Token usingToken)
             if (itemAst == null)
             {
                 ReportError(itemToken.Extent, () => ParserStrings.InvalidValueForUsingItemName, itemToken.Text);
+                // ErrorRecovery: If there is no identifier, skip whole 'using' line
+                SyncOnError(true, TokenKind.NewLine);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: Please clarify. Example?

File: test/powershell/Language/Classes/scripting.Classes.using.tests.ps1 (hunk 3 above)
Comment (iSazonov) [inline test/powershell/Language/Classes/scripting.Classes.using.tests.ps1]: Fixed.

File: src/System.Management.Automation/engine/parser/Parser.cs (hunk 4 above)
Comment (lzybkr) [inline src/System.Management.Automation/engine/parser/Parser.cs]: `SyncOnError` skips tokens until it finds the one(s) you say are where to stop. Semicolon terminates a using statement in the same way a newline does, so we should assume we saw the end of a using statement if we see a semicolon or newline.

File: test/powershell/Language/Classes/scripting.Classes.using.tests.ps1 [hunk 5]
```diff
@@ -261,6 +261,30 @@ New-Object Foo
...
 
+        It "report an error when 'using module' terminating by NewLine" {
+            $err = Get-ParseResults "using module"
+            $err.Count | Should Be 1
+            $err[0].ErrorId | Should Be 'MissingUsingItemName'
+        }
+
+        It "report an error when 'using module' terminating by Semicolon" {
+            $err = Get-ParseResults "using module; $testvar=1"
+Write-Host $err
```
Comment (lzybkr) [inline test/powershell/Language/Classes/scripting.Classes.using.tests.ps1]: Debugging output should be removed

Comment (iSazonov) [inline test/powershell/Language/Classes/scripting.Classes.using.tests.ps1]: Fixed.

File: src/System.Management.Automation/engine/parser/Parser.cs (hunk 4 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: Sorry, I'm not woke up today :-) and mixed a comma and semicolon.
Fixed.

**Review by lzybkr — APPROVED** [review]

**Pull request merged**